### DISCUSSION TITLE
Vitest threads

### DISCUSSION THREAD

**Discussion by zaalbarxx** [vitest-dev/vitest#4088] [Q&A]
Hi!
Today I stumbled upon an issue with running Vitest with JSDOM. Long story short JSDOM tries to automatically require `canvas` lib if it detects it exists. Since I am also using Fabric.js which has `canvas` as optional dependency it also gets installed. Now, after some digging I found this issue:
[https://github.com/vitest-dev/vitest/issues/740](url)
and some kind of fix for that (or at least from what I understand):
[https://github.com/vitest-dev/vitest/pull/2772](url)
Now the question is - is running Vitest with `threads: false` cause it to run in single thread ? How does Jest handle this stuff since previously in our project we were using Jest and it worked well with parallelization ? Just curious how could I tackle this issue. From what I've read it is not really possible to mock canvas (that should be sufficient since I don't really need to "run" canvas in tests) ?
I'd just want to get a better understanding of this and what are the options here besides of `threads:false` which I assume hit performance a lot (just like Jest's `--runInBand`).

**Comment by AriPerkkio** [comment]
> Now the question is - is running Vitest with `threads: false` cause it to run in single thread ? 

Yes, it runs in a single process. Once https://github.com/vitest-dev/vitest/pull/3925 is finished Vitest will be able to run multiple `child_process` in parallel. I think there have been reports that `canvas` crashes a lot when run in `worker_threads` so that PR should improve speed for those cases.

> How does Jest handle this stuff since previously in our project we were using Jest and it worked well with parallelization ?

Jest uses `child_process` by default. They provide an option for enabling threads: https://jestjs.io/docs/configuration#workerthreads. You would likely run into same options if `workerThreads` was enabled.
